lib/Models/CatalogItem.js

```diff
@@ -1167,4 +1283,104 @@ function isShownChanged(catalogItem) {
 }
 
+/**
+ * This function adds / removes a subscription to copy the time from the terria clock into the items clock when .useOwnClock is false.
+ *
+ * @param catalogItem The item to add / remove the terria clock watching subscription from.
+ * @private
+ */
+function updateTerriaClockWatcher(catalogItem) {
+    if (catalogItem._mirroringTerriaClock && !defined(catalogItem._removeTerriaClockWatch)) {
+        // We are using the terria clock, add a subscription to copy the current time into this items clock.
+        catalogItem._removeTerriaClockWatch = catalogItem.terria.clock.onTick.addEventListener(terriaClock => {
+            if (!JulianDate.equals(terriaClock.currentTime, catalogItem.clock.currentTime)) {
+                updateCurrentTime(catalogItem, terriaClock.currentTime);
+            }
+        });
+
+        updateCurrentTime(catalogItem, catalogItem.terria.clock.currentTime);
+    } else if (!catalogItem._mirroringTerriaClock && defined(catalogItem._removeTerriaClockWatch)) {
+        // We are no longer using the terria clock, remove the subscription.
+        catalogItem._removeTerriaClockWatch();
+        catalogItem._removeTerriaClockWatch = undefined;
+    }
+}
+
+/**
+ * Updates the item.clock.currentTime using the value provided.
+ * This function does this so that all clients subscribed to the items clock with DataSourceClock.definitionChanged are notified.
+ *
+ * @param catalogItem The CatalogItem to set the current time for.
+ * @param updatedTime The new current time to use.
+ * @private
+ */
+function updateCurrentTime(catalogItem, updatedTime) {
+    if (defined(catalogItem.clock)) {
+        catalogItem.clock.currentTime = JulianDate.clone(updatedTime);
+    }
+}
+
+/**
+ * Syncs the time between the terria.clock and the catalogItem.clock.
+ *
+ * The sync is performed in the direction that is required depending on the state change true->false / false->true of useOwnClock.
+ *
+ * @param catalogItem The CatalogItem to sync the time for.
+ * @private
+ */
+function useOwnClockChanged(catalogItem) {
+    // If we are changing the state, copy the time from the clock that was in use to the clock that will be in use.
+    if (catalogItem._lastUseOwnClock !== catalogItem.useOwnClock) {
+        if (catalogItem.useOwnClock) {
+            // This is probably not needed, since we should be doing it on update, but just do this here explicitly
+            // to be sure it is immediately current before we change.
+            updateCurrentTime(catalogItem, catalogItem.terria.clock.currentTime);
+        } else {
+            catalogItem.terria.clock.currentTime = JulianDate.clone(catalogItem.clock.currentTime);
+        }
+    }
+
+    catalogItem._lastUseOwnClock = catalogItem.useOwnClock;
+}
+
+function timeAtIndex(catalogItem, index) {
+    if (defined(index) && defined(catalogItem.availableDates) && (index >= 0) && (index < catalogItem.availableDates.length)) {
+        return catalogItem.availableDates[index];
```

Comment:
According to the docs for the `availableDates` property, elements in `availableDates` can be either a Date or an interval.  I'm not sure offhand if it's true.

Reply:
Types would make it much easier to be sure. Looking at the code this doesn't seem to be the case:
```
                for (let i = 0; i < this.intervals.length; i++) {
                    datetimes.push(JulianDate.toDate(this.intervals.get(i).start, 3));
                }
```
https://github.com/TerriaJS/terriajs/blob/5212564af3c38f09e4e8327f241ff2849644be30/lib/Models/ImageryLayerCatalogItem.js#L187
https://github.com/TerriaJS/terriajs/blob/5212564af3c38f09e4e8327f241ff2849644be30/lib/Models/TableCatalogItem.js#L216

But this line has somewhat bothered me anyway since it was just refactored into that from the original line that it replaced, (https://github.com/TerriaJS/terriajs/pull/2819/files#diff-a8595cd3f06d548e02a70b13e4df2e6cL69, https://github.com/TerriaJS/terriajs/pull/2819/files#diff-a8595cd3f06d548e02a70b13e4df2e6cL74 and so on), but has bothered me since its not really symmetric where the value comes from. Have updated to fix the code now that I've read the docco around available dates since there shouldn't be much of a performance difference here anyway and this more symmetric implementation seems cleaner.

5212564af3c38f09e4e8327f241ff2849644be30